json-schemas/HOL-PART4.md

```diff
@@ -0,0 +1,166 @@
+# Part 4: Schema Dependencies
+
+Another interesting feature offered in [JSON Schema Validation](https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.4.5)
+is the utilization of the `dependencies` keyword to allow the schema to be
+defined based on specific properties. This can be of great benefit to
+move application logic natively into the database. These dependencies
+come in two forms:
+
++ **Property dependencies** state that if a specified property is present,
+other properties must be present as well.
++ **Schema dependencies** define a *change* in the schema when a given
+property is present.
+
+### Property Dependencies
+
+1. For the case of property dependencies let's suppose our schema represents
+students and if they have graduated from a program. If they have graduated
+we want to be able to stay in touch with them and need their mailing address.
+
+```
+db.students.drop()
+db.createCollection ( "students",
+{
+    validator:
+    {
+        $jsonSchema:
+      {
+        bsonType: "object",
+        required: ["name"],
+        properties:
+        {
+            _id: {},
+            name: {
+                bsonType: ["string"],
+                description: "'name' is a required string"
+            },
+            graduated: {
+                bsonType: ["bool"],
+                description: "'graduated' is an optional boolean value"
+            },
+            mailing_address: {
+                bsonType: ["string"],
+                description: "'mailing_address' is an optional string value"
+            }
+        },
+        dependencies: {
+            "graduated": ["mailing_address"]
+        }
+      }
+    }
+})
+```
+2. We can test our schema validation rules:
+```
+db.students.insertOne({name: "Alena Weber",
+graduated: true,
+mailing_address: "123 Main Street"}) // works
+
+db.students.insertOne({name: "Jamal Aattou",
+graduated: true}) // doesn't work, no mailing_address
+
+db.students.insertOne({name: "Shannon Bradshaw"}) // works
+
+db.students.insertOne({name: "Andrei Popov",
+mailing_address: "456 Oak Street"}) // works, having an address without graduation is fine, dependencies are not bidirectional
+```
```

Comment:
What happens if graduated = false? Doesn't that make *mailing_address* required too? If so, maybe you should point that out and mention that the assumption is that the app doesn't add the `graduated` attribute unless the student has indeed graduated.